v2/tools/generator/internal/astmodel/readonly_object_type.go

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+// ReadonlyObjectType allows exposing an ObjectType for querying while prohibiting any modification.
+// This is useful as a convenience to consumers for working with ObjectTypes, while ensuring they
+// make any modifications using a TypeVisitor (or one of the Injectors based on TypeVisitor).
+type ReadonlyObjectType interface {
+	// Properties returns all the property definitions
+	Properties() ReadOnlyPropertySet
+
+	// Property returns the details of a specific property based on its unique case-sensitive name
+	Property(name PropertyName) (*PropertyDefinition, bool)
```

Comment:
Are we safe from the `PropertyDefinition` being modified here?

Maybe I can tell when I look at usage.

Reply:
Each `PropertyDefinition` is immutable, but yes, a new one could be created here.

That's not a problem because the new one would be created from the _whole_ existing property definition. The problem I'm solving there is that the `SpecType` previously exposed in the result struct was incomplete (any wrapper types already stripped) so using that as a base for further modification was flawed.